src/directory/managed_directory.rs

```diff
@@ -207,19 +210,198 @@ impl ManagedDirectory {
+        }
+    }
+
+    pub fn index_version(&self) -> u8 {
+        match self {
+            Footer::V0(_) => 0,
+        }
+    }
+
+    pub fn tantivy_version(&self) -> (u8, u8, u8) {
+        match self {
+            Footer::V0(footer) => footer.tantivy_version,
+        }
+    }
+
+    pub fn crc(&self) -> u32 {
+        match self {
+            Footer::V0(footer) => footer.crc,
+        }
+    }
+
+    pub fn size(&self) -> usize {
+        match self {
+            Footer::V0(_) => V0::size() as usize,
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct V0 {
+    pub tantivy_version: (u8, u8, u8),
+    pub crc: u32,
+}
+
+impl V0 {
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut res = vec![0; 10];
+        res[0] = 0;
+        res[1] = self.tantivy_version.0;
+        res[2] = self.tantivy_version.1;
+        res[3] = self.tantivy_version.2;
+        LittleEndian::write_u32(&mut res[4..8], self.crc);
+        LittleEndian::write_u16(&mut res[8..], Self::size());
+        res
+    }
+
+    pub fn from_bytes(footer: &[u8]) -> Self {
+        assert_eq!(footer[0], 0);
+        assert_eq!(LittleEndian::read_u16(&footer[8..10]), 10);
+        let tantivy_version = (footer[1], footer[2], footer[3]);
+        let crc = LittleEndian::read_u32(&footer[4..8]);
+
+        V0 {
+            tantivy_version,
+            crc,
+        }
+    }
+
+    pub fn from_crc(crc: u32) -> Self {
+        Self {
+            tantivy_version: (
+                env!("CARGO_PKG_VERSION_MAJOR").parse().unwrap(),
+                env!("CARGO_PKG_VERSION_MINOR").parse().unwrap(),
+                env!("CARGO_PKG_VERSION_PATCH").parse().unwrap(),
+            ),
+            crc,
+        }
+    }
+    pub fn size() -> u16 {
+        10
+    }
+}
+
+struct FooterProxy<W: TerminatingWrite> {
+    /// always Some except after terminate call
+    hasher: Option<Hasher>,
+    /// always Some except after terminate call
+    writer: Option<W>,
+}
+
+impl<W: TerminatingWrite> FooterProxy<W> {
+    fn new(writer: W) -> Self {
+        FooterProxy {
+            hasher: Some(Hasher::new()),
+            writer: Some(writer),
+        }
+    }
+}
+
+impl<W: TerminatingWrite> Write for FooterProxy<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let count = self.writer.as_mut().unwrap().write(buf)?;
+        self.hasher.as_mut().unwrap().update(&buf[..count]);
+        Ok(count)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        self.writer.as_mut().unwrap().flush()
+    }
+}
+
+impl<W: TerminatingWrite> TerminatingWrite for FooterProxy<W> {
+    fn terminate_ref(&mut self, _: AntiCallToken) -> io::Result<()> {
+        let crc = self.hasher.take().unwrap().finalize();
+        let footer = Footer::V0(V0::from_crc(crc)).to_bytes();
+        let mut writer = self.writer.take().unwrap();
+        writer.write_all(&footer)?;
+        writer.flush()?; //should we assume calling terminate on inner already flush?
```

Comment:
I'm ok with having flush in `terminate`.

Reply:
You can remove this comment and comment explicitely in the `TerminatingWrite` that `.terminate_ref()` should flush.